### PR TITLE
Python 3.12: Refactor out Distutils

### LINKED ISSUE #2821: Refactor out Distutils
[distutils is deprecated](https://www.python.org/dev/peps/pep-0632/) (PEP 632).

We need to refactor it out for newer modules such as:
- setuptools (already used, use more functions from it)
- [packaging](https://github.com/pypa/packaging)
- [build](https://github.com/pypa/build)

Upstream example from pybind11:
- https://github.com/pybind/cmake_example/blob/master/setup.py

### PR DESCRIPTION
**Pull request by ax3l**
Fix #2821 for Python 3.12

Refs.:
- https://peps.python.org/pep-0632/#migration-advice
- https://github.com/AMReX-Codes/pyamrex/pull/180

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -9,6 +6,7 @@
 import sys
 
 from setuptools import Extension, setup
+import setuptools.command.build
```
[the remark this answers is not in the document]
Comment (ax3l) [reply setup.py]: outdated

File: .github/workflows/cuda.yml [hunk 2]
```diff
@@ -18,16 +18,12 @@ jobs:
...
-      #   https://github.com/mpi4py/mpi4py/pull/159
-      #   https://github.com/mpi4py/mpi4py/issues/157#issuecomment-1001022274
-      SETUPTOOLS_USE_DISTUTILS: stdlib
     steps:
     - uses: actions/checkout@v3
     - uses: actions/setup-python@v4
       name: Install Python
       with:
-        python-version: '3.10'
+        python-version: '3.x'
```
Comment (dpgrote) [inline .github/workflows/cuda.yml:26]: What does this do?

Comment (ax3l) [reply .github/workflows/cuda.yml:26]: Unconstrain the upper Python version that gets installed. I constrained it in https://github.com/ECP-WarpX/WarpX/commit/ed13486ac0495b049958989022a9c76e8ef8f042 to unbreak CI by avoiding Python 3.12 for a few days.

I want it unconstrained again, so I see if there are breaking changes once new Python versions are released.

Ref.: https://github.com/actions/setup-python#setup-python

File: Tools/machines/lassen-llnl/install_v100_ml.sh [hunk 3]
```diff
@@ -64,6 +64,7 @@ rm -rf build
 cd -
 
 # optional: optimas dependencies (based on libEnsemble & ax->botorch->gpytorch->pytorch)
+TODO: scikit-learn needs a BLAS hint
```
Comment (roelof-groenewald) [inline Tools/machines/lassen-llnl/install_v100_ml.sh]: Is this okay to not be a comment?
```suggestion
# TODO: scikit-learn needs a BLAS hint
```

**Review by roelof-groenewald — APPROVED** [review]
Looks good. Just one minor comment.

File: Tools/machines/lassen-llnl/install_v100_ml.sh (hunk 3 above)
Comment (ax3l) [reply Tools/machines/lassen-llnl/install_v100_ml.sh]: Oh no, added accidentally

**Pull request merged**